File: archive/p/python/merge_sort.py
import sys


def merge_sort(xs):
    def sort(xs):
        if len(xs) <= 0:
            return []
        if len(xs) == 1:
            return xs
        return sort([merge(xs[0], xs[1])] + sort(xs[2:]))
    split_xs = [[x] for x in xs]
    return sort(split_xs)[0]


def merge(xs, ys):
    if len(xs) <= 0:
        return ys
    if len(ys) <= 0:
        return xs
    if xs[0] < ys[0]:
        return [xs[0]] + merge(xs[1:], ys)
    return [ys[0]] + merge(xs, ys[1:])


def input_list(list_str):
    return [int(x.strip(" "), 10) for x in list_str.split(',')]


def exit_with_error():
    print('Usage: please provide a list of at least two integers to sort in the format "1, 2, 3, 4, 5"')
    sys.exit(1)


def main(args):
    try:
        xs = input_list(args[0])
        if len(xs) <= 1:
            exit_with_error()
        print(merge_sort(xs))
    except (IndexError, ValueError):
        exit_with_error()


if __name__ == "__main__":
    main(sys.argv[1:])

File: archive/p/perl/merge-sort.pl
#!/usr/bin/env perl
use v5.42;

use feature qw/keyword_any/;
no warnings 'experimental::keyword_any';

sub usage {
    say 'Usage: please provide a list of at least two integers to sort in the format "1, 2, 3, 4, 5"';
    exit;
}

sub parse_list ($s) {
    return undef unless defined $s;

    my @vals = split /\s*,\s*/, $s;

    return undef if @vals < 2;
    return undef if any { $_ !~ /\A-?\d+\z/ } @vals;

    return [ map 0 + $_, @vals ];
}

# Note: Perl's `sort` is implemented via merge sort as of v5.42, so one could use that instead.

sub merge_sort ($a) {
    return $a if @$a <= 1;

    my $mid = @$a >> 1;

    my $left  = [ @{$a}[ 0 .. $mid - 1 ] ];
    my $right = [ @{$a}[ $mid .. $#$a ] ];

    merge_sort($left);
    merge_sort($right);

    @$a = _merge( $left, $right );
    return $a;
}

sub _merge ( $left, $right ) {
    my @out;

    while ( @$left && @$right ) {
        push @out, $left->[0] <= $right->[0]
          ? shift @$left
          : shift @$right;
    }

    push @out, @$left, @$right;

    return @out;
}

my ($input) = @ARGV;
my $a = parse_list($input) or usage();

merge_sort($a);
say join ', ', @$a;
